src/common.cc

```diff
@@ -111,4 +121,337 @@ Join(const std::vector<std::string>& vec, const std::string& delim)
+      while (next_offset < byte_size) {
+        uint32_t* length = reinterpret_cast<uint32_t*>(base + next_offset);
+        *length = bswap_length(*length, host_byte_order, &host_length);
+        next_offset += sizeof(uint32_t) + host_length;
+      }
+      break;
+    }
+    default: {
+    }
+  }
+}
+
+/// Swaps byte order in place with support for non-contiguous arrays. For
+/// non-contiguous arrays the initial call must pass an empty partial result and
+/// 0 for the initial next_offset. Subsequent calls should pass the returned
+/// values without modification.
+///
+/// \param[in] datatype data type of array
+/// \param[in, out] base pointer to array of data type elements to convert
+/// \param[in] byte_size size of array in bytes
+/// \param[in] host_byte_order whether original array is in host byte
+/// order
+/// \param[in, out] partial_result vector to store partial results
+/// \param[in, out] next_offset intermediate value used for storing next offset
+/// for BYTES data type
+///
+void
+SwapEndian(
+    TRITONSERVER_DataType datatype, char* base, size_t byte_size,
+    bool host_byte_order, std::vector<char*>& partial_result,
+    size_t& next_offset)
+{
+  switch (datatype) {
+    case TRITONSERVER_TYPE_UINT16:
+    case TRITONSERVER_TYPE_INT16:
+    case TRITONSERVER_TYPE_FP16: {
+      SwapEndian<uint16_t>(base, byte_size, partial_result);
+      break;
+    }
+    case TRITONSERVER_TYPE_UINT32:
+    case TRITONSERVER_TYPE_INT32:
+    case TRITONSERVER_TYPE_FP32: {
+      SwapEndian<uint32_t>(base, byte_size, partial_result);
+      break;
+    }
+    case TRITONSERVER_TYPE_UINT64:
+    case TRITONSERVER_TYPE_INT64:
+    case TRITONSERVER_TYPE_FP64: {
+      SwapEndian<uint64_t>(base, byte_size, partial_result);
+      break;
+    }
+    case TRITONSERVER_TYPE_BYTES: {
+      uint32_t host_length = 0;
+      while (next_offset < byte_size) {
+        size_t partial_offset = partial_result.size();
+        bswap_length_partial(
+            base + next_offset, byte_size - next_offset, host_byte_order,
+            &host_length, partial_result);
+        if (partial_result.size() == 0) {
+          next_offset += sizeof(uint32_t) - partial_offset + host_length;
+        } else {
+          next_offset = 0;
+          return;
+        }
+      }
+      next_offset -= byte_size;
+      break;
+    }
+    default: {
+    }
+  }
+}
+
+void
+HostToLittleEndian(TRITONSERVER_DataType datatype, char* base, size_t byte_size)
+{
+  SwapEndian(datatype, base, byte_size, true);
+}
+
+void
+LittleEndianToHost(TRITONSERVER_DataType datatype, char* base, size_t byte_size)
+{
+  SwapEndian(datatype, base, byte_size, false);
+}
+
+void
+LittleEndianToHost(
+    TRITONSERVER_DataType datatype, char* base, size_t byte_size,
+    std::vector<char*>& partial_result, size_t& next_offset)
+{
+  SwapEndian(datatype, base, byte_size, false, partial_result, next_offset);
+}
+
+TRITONSERVER_DataType
+GetDataTypeForRawInput(
```

Comment:
Was on the fence here - I opted for an error at the call site for when it wasn't able to determine the data type. If there is an issue with the model in terms of being able to send a raw binary input - that is also caught by other checks in normalize, etc. In this case I think having the more general error "couldn't determine data type" was sufficient. The only real check here is that there is an input (otherwise model, model version, etc. should be caught already).

Let me know if you feel otherwise -

Reply:
I think it would be nicer if we can return the error. We can still keep the message as "couldn't determine data type". I agree that these calls would most likely not return any error but the logic on why the error checking is not required (i.e., normalize is first called) is not apparent in the first glance.